packages/core/src/model.ts

```diff
@@ -0,0 +1,38 @@
+import { assign } from './actions';
+import type {
+  AssignAction,
+  Assigner,
+  PropertyAssigner,
+  ExtractEvent,
+  EventObject
+} from './types';
+
+export interface ContextModel<TContext, TEvent extends EventObject> {
+  initialContext: TContext;
+  assign: <TEventType extends TEvent['type'] = TEvent['type']>(
+    assigner:
+      | Assigner<TContext, ExtractEvent<TEvent, TEventType>>
+      | PropertyAssigner<TContext, ExtractEvent<TEvent, TEventType>>,
+    eventType?: TEventType
+  ) => AssignAction<TContext, ExtractEvent<TEvent, TEventType>>;
+  reset: () => AssignAction<TContext, any>;
+}
+
+export function createModel<TContext, TEvent extends EventObject>(
+  initialContext: TContext
+): ContextModel<TContext, TEvent> {
+  const model: ContextModel<TContext, TEvent> = {
+    initialContext,
+    assign,
+    reset: () => assign(initialContext)
+  };
+
+  return model;
+}
+
+export function assertEvent<
```

Comment:
This name is somewhat misleading because it leads the user to believe that this is an assertion function (`asserts event is TEvent`) but this is a type predicate (`event is TEvent`).

Besides that - could you add a test showcasing how this is supposed to be used?

Reply:
Removed